### PR TITLE
Feat/task 03

### PR DESCRIPTION
**Pull request by numatay**
Closes Task-03

Main tasks:
1 - product-service serverless.yml contains configuration for 2 lambda functions, API is not working at all, but YAML configuration is correct - DONE

2 - The getProductsList OR getProductsById lambda function returns a correct response (POINT1) - **DONE**
The URL for getProductsList is https://p6cq8jgiid.execute-api.eu-west-1.amazonaws.com/products

3 - The getProductsById AND getProductsList lambda functions return a correct response code (POINT2) - **DONE**
The URL for getProductById is https://pnes1azyil.execute-api.eu-west-1.amazonaws.com/dev/products/5b718fac-8a4c-48b5-9379-215f1526598c

4 - Your own Frontend application is integrated with product service (/products API) and products from product-service are represented on Frontend. AND POINT1 and POINT2 are done. - **DONE**
The link to PR on frontend repo (rolling-scopes-school/nodejs-aws-fe#134)
The deployed version of the web app with backend integration is here: https://d5bzm0spt9d1j.cloudfront.net

Additional (optional) tasks (but nice to have):
+1 - Async/await is used in lambda functions - **DONE**
+1 - ES6 modules are used for product-service implementation - **DONE**
+1 - Webpack is configured for product-service - **DONE**
+1 - SWAGGER documentation is created for product-service - **NOT_DONE**
+1 - Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered) (You may use JEST) - **NOT_DONE**
+1 - Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase. - **DONE**
+1 - Main error scenarios are handled by API ("Product not found" error, try-catch blocks are used in lambda handlers). - **DONE**

### CODE REVIEW

File: product-service/package.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "product-service",
+  "version": "1.0.0",
+  "description": "Serverless webpack example using Typescript",
+  "main": "handler.js",
+  "scripts": {
+    "deploy": "serverless deploy -s dev -r eu-west-1 -v",
```
Comment (LoraBard) [inline product-service/package.json]: You can move region into serverless.ts  file (provider)

File: product-service/serverless.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      ]
+    },
+    getProductById: {
+      handler: 'handler.getProductById',
+      events: [
+        {
+          http: {
+            method: 'get',
+            path: 'products/{productId}',
+            cors: true,
```
Comment (LoraBard) [inline product-service/serverless.ts]: or  
provider 
  httpApi:
    cors: true
And use httpApi instead of http

File: product-service/src/db.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    imageURL: string;
+    price: string;
+}
+
+class DatabaseServiceMock {
+    products: Map<string, Product>;
+
+    constructor() {
+        this.products = new Map<string, Product>();
+        Array.from(productsList.keys()).forEach(productKey => {
```
Comment (LoraBard) [inline product-service/src/db.ts:14]: Possible. It looks a bit complicated. Did you think about complixety of this solution? Just:
private products = productsList;
getProducts(): Product[] {
return this.products;
}
getProductsById(id: string):Product {
  const product = this.products.find((item:Product) => item.id === id);
  ...
}

File: product-service/src/db.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            const productItem = productsList[productKey];
+            this.products.set(productItem.id, {
+                id: productItem.id,
+                imageURL: productItem.imageURL,
+                price: productItem.price,
+            });
+        });
+    }
+
+    async getProducts() {
```
Comment (LoraBard) [inline product-service/src/db.ts]: Why async (without await)? Just use Promise.resolve() or create response handler function

File: product-service/src/getAllProducts.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+export const getAllProducts = async (_event, _context) => {
+    const products = await db.getProducts();
+    return {
+        statusCode: 200,
+        headers: {
+            'Access-Control-Allow-Origin': '*',
+            'Access-Control-Allow-Credentials': true,
+        },
+        body: JSON.stringify(products)
```
Comment (LoraBard) [inline product-service/src/getAllProducts.ts]: Duplicate of code. Please move to separate function

File: product-service/src/getProductById.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import db from "./db";
+
+export const getProductById = async (event, _context) => {
+
+    const { pathParameters } = event;
+    console.log("Event:", event, pathParameters);
```
Comment (LoraBard) [inline product-service/src/getProductById.ts]: remove console.log

**Review by LoraBard — CHANGES_REQUESTED** [review]

File: product-service/src/getAllProducts.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import db from './db';
+import 'source-map-support/register';
+
+export const getAllProducts = async (_event, _context) => {
```
Comment (LoraBard) [inline product-service/src/getAllProducts.ts]: Please add types

File: product-service/src/getProductById.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import db from "./db";
+
+export const getProductById = async (event, _context) => {
```
Comment (LoraBard) [inline product-service/src/getProductById.ts]: the same

**Review by LoraBard — CHANGES_REQUESTED** [review]

File: product-service/package.json (hunk 1 above)
Comment (numatay) [reply product-service/package.json]: Moved to `serverless.ts` and cleaned up `npm run deploy` script

File: product-service/serverless.ts (hunk 2 above)
Comment (numatay) [reply product-service/serverless.ts]: Moved `cors: true` to `httpApi` and reused `httpApi` with function

File: product-service/src/db.ts (hunk 4 above)
Comment (numatay) [reply product-service/src/db.ts]: Removed `async` keyword from DB service and changed return value with `Promise.resolved` wrapped.

File: product-service/src/getAllProducts.ts (hunk 5 above)
Comment (numatay) [reply product-service/src/getAllProducts.ts]: Extracted to `addCorsHeaders` utilities and reused it in both functions.

File: product-service/src/getProductById.ts (hunk 6 above)
Comment (numatay) [reply product-service/src/getProductById.ts]: Cleaned up code.

File: product-service/src/getAllProducts.ts (hunk 7 above)
Comment (numatay) [reply product-service/src/getAllProducts.ts]: Added return types for the functions (`APIGatewayProxyHandler` from `aws-lambda` library)

File: product-service/src/getProductById.ts (hunk 8 above)
Comment (numatay) [reply product-service/src/getProductById.ts]: Added return types for the functions (`APIGatewayProxyHandler` from `aws-lambda` library)

File: product-service/src/db.ts (hunk 3 above)
Comment (numatay) [reply product-service/src/db.ts:14]: Although this code looks confusing, I want to use a `map`  for the underlying data structures.
I would like to keep it this way if it doesn't affect the grading. ;)

Comment (LoraBard) [reply product-service/src/db.ts:14]: @numatay Of course, it doesn't affect the grading

**Comment by alexeyMohnatkin** [comment]
getProductById responds with 404 for any id

**Comment by numatay** [comment]
> getProductById responds with 404 for any id

I must have messed up when deployed and updated the stack. Will fix it in a moment!

@alexeyMohnatkin, I updated the URL for `getProductById` in the description. Please have a look at it.

**Pull request merged**